examples/utils/example-utils/src/migrator/sameContainerMigrator.ts

```diff
@@ -73,221 +81,175 @@ export class SameContainerMigrator
+			return;
+		}
+
+		this._detachedModel = await this.modelLoader.createDetached(this._acceptedVersion);
+		this._migratedModel = this._detachedModel.model;
 	};
 
-	private readonly ensureMigrating = () => {
-		// ensureMigrating() is called when we reach the "migrating" state. This should likely only happen once, but
-		// can happen multiple times if we disconnect during the migration process.
+	private readonly getExportedData = async () => {
+		const acceptedSeqNum = this.currentModel.migrationTool.acceptedSeqNum;
+		assert(acceptedSeqNum !== undefined, "acceptedSeqNum should be defined");
+		this._pausedModel = await this.modelLoader.loadExistingPaused(
+			this._currentModelId,
+			acceptedSeqNum,
+		);
+		assert(
+			this.pausedModel.container.deltaManager.lastSequenceNumber === acceptedSeqNum,
+			"paused model should be at accepted sequence number",
+		);
+		this._exportedData = await this.pausedModel.exportData();
+	};
 
-		if (!this.connected) {
-			// If we are not connected we should wait until we reconnect and try again. Note: we re-enter the state
-			// machine, since its possible another client has already completed the migration by the time we reconnect.
-			this.currentModel.once("connected", this.takeAppropriateActionForCurrentMigratable);
+	private readonly generateTransformedData = async () => {
+		// TODO: Is there a reasonable way to validate at proposal time whether we'll be able to get the
+		// exported data into a format that the new model can import?  If we can determine it early, then
+		// clients with old ModelLoaders can use that opportunity to dispose early and try to get new
+		// ModelLoaders.
+		assert(this._migratedModel !== undefined, "this._migratedModel should be defined");
+		if (this._migratedModel.supportsDataFormat(this._exportedData)) {
+			// If the migrated model already supports the data format, go ahead with the migration.
+			this._transformedData = this._exportedData;
+		} else if (this.dataTransformationCallback !== undefined) {
+			// Otherwise, try using the dataTransformationCallback if provided to get the exported data into
+			// a format that we can import.
+			try {
+				this._transformedData = await this.dataTransformationCallback(
+					this._exportedData,
+					this._migratedModel.version,
+				);
+			} catch {
+				// TODO: This implies that the contract is to throw if the data can't be transformed, which
+				// isn't great.  How should the dataTransformationCallback indicate failure?
+				this.emit("migrationNotSupported", this._acceptedVersion);
+				return;
+			}
+		} else {
+			// We can't get the data into a format that we can import, give up.
+			this.emit("migrationNotSupported", this._acceptedVersion);
 			return;
 		}
+		await this._migratedModel.importData(this._transformedData);
 
-		if (this._migrationP !== undefined) {
-			return;
-		}
+		// Store the detached model for later use and retry scenarios
+		this._preparedDetachedModel = this._detachedModel;
```

Comment:
This part can split out and live next to the detached model creation (see comments lower down about creating the detached container after transform finishes)